backfiller/backfiller/main.py

```diff
@@ -129,159 +114,308 @@ def get_remote_segment(base_dir, node, stream, variant, hour, missing_segment,
+				self.start_worker(node)
+			to_stop = exisiting_nodes - new_nodes
+			for node in to_stop:
+				self.stop_worker(node)
+
+			if self.run_once:
+				break
+
+			self.stopping.wait(common.jitter(self.NODE_INTERVAL))
 
-def backfill_node(base_dir, node, stream, variants, hours=None, segment_order='random', recent_cutoff=60):
-	"""Backfill from remote node.
+		else:
+			self.stop()
 
-	Backfill from node/stream/variants to base_dir/stream/variants.
+		for worker in self.workers.values():
+			worker.done.wait()
 
-	Keyword arguments:
-	hours -- If None (default), backfill all available hours. If iterable,
-		backfill only hours in iterable. Otherwise backfill the last N hours,
-		starting with the lastest.
-	segment_order -- If 'random', randomise the order of segments (default).
-		If 'forward', sort the segment in ascending order. If 'reverse', sort
-		the segments in descending order. Otherwise, do not change the order of
-		segments.
-		use start and stop to limit which segments are backfilled.
-	recent_cutoff -- Skip backfilling segments younger than this number of
-		seconds to prioritise letting the downloader grab these segments."""
-
-	logging.info('Starting backfilling from {}'.format(node))
-
-	if hours is None:
-		# gather all available hours from all variants and take the union
-		hours = set().union(*[
-			list_remote_hours(node, stream, variant)
-			for variant in variants
-		])
-	elif is_iterable(hours):
-		hours = list(hours) # coerce to list so it can be sorted
-	else:
-		n_hours = hours
-		if n_hours < 1:
-			raise ValueError('Number of hours has to be 1 or greater')
-		now = datetime.datetime.utcnow()
-		hours = [(now - i * datetime.timedelta(hours=1)).strftime(HOUR_FMT) for i in range(n_hours)]
-
-	for variant in variants:
-
-		for hour in hours:
-			logging.info('Backfilling {}/{}/{}'.format(stream, variant, hour))
-
-			local_segments = set(list_local_segments(base_dir, stream, variant, hour))
-			remote_segments = set(list_remote_segments(node, stream, variant, hour))
-			missing_segments = list(remote_segments - local_segments)
-
-			# useful if running in parallel so multiple nodes don't request the same segment at the same time
-			if segment_order == 'random':
-				random.shuffle(missing_segments)
-			elif segment_order == 'forward':
-				missing_segments.sort()
-			elif segment_order == 'reverse':
-				missing_segments.sort(reverse=True)
-
-			for missing_segment in missing_segments:
+	def get_nodes(self):
+		"""List address of other wubloaders.
+		
+		This returns a list of the other wubloaders as strings of the form
+		'protocol://host:port/'"""
+		nodes = self.static_nodes + []
+	
+		if self.node_file is not None:
+			self.logger.info('Fetching list of nodes from {}'.format(self.node_file))
+			try:
+				with open(self.node_file) as f:
+					for line in f.readlines():
+						substrs = line.split()
+						if not len(line) or substrs[0][0] == '#':
+							continue
+						nodes.append(substrs[0])
+			except IOError:
```

Comment:
Rather than handling errors in file or database lookups as ignoring their contents (which will cause the respective workers to be shut down), I think errors in get_nodes should simply be propagated upwards and handled in `run()` with a retry. That way if eg. the file is unavailable briefly, or the database is down, we maintain our existing set of workers instead of stopping everything.

Reply:
Ok, I've copied the same back off and number or tries from the workers to handle errors in get_nodes. If max retries is reached, existing workers are allow to keep running but no further attempts are made to call get_nodes.